Develop

kid-sid opened the pull request:
Added rate limiter for all the APIs using a centralized rate limiting service.

kid-sid committed:
Removed outdated test file status from PROJECT_STATUS.md and noted missing tests for prompt history, feedback system, rate limiting, and guardrails. This update streamlines the document to focus on current testing coverage and highlights areas needing attention.

kid-sid committed:
Added project_status.md to .gitignore to prevent tracking of project status documentation in version control.

kid-sid committed:
Updated .gitignore to change the entry for project_status.md to PROJECT_STATUS.md, ensuring consistent casing for the file name.

kid-sid committed:
Added rate limiting middleware to main application and implement status endpoint for rate limiting service. Bypass specified endpoints for rate limiting checks to ensure essential services remain accessible.

kid-sid committed:
Added rate limits configuration for various endpoints and user tiers. This new file centralizes rate limiting settings, including request limits and burst capacities, to facilitate easier adjustments and management of API usage across different service tiers.

kid-sid committed:
Added __init__.py to middleware package to establish a namespace for middleware components.

kid-sid committed:
Added FastAPI Rate Limiting Middleware

This commit introduces a new middleware for FastAPI that automatically applies rate limiting to all API endpoints. The middleware includes features such as user-based and IP-based limiting, configurable bypass for certain endpoints, and proper HTTP headers for rate limiting information. It also provides a factory function for custom configuration of the middleware.

kid-sid committed:
Implemented a new endpoint to retrieve the rate limiting status for inference endpoints. This addition allows authenticated users to check their current rate limit usage, limits, and usage percentages, enhancing the API's usability and transparency regarding rate limits.

kid-sid committed:
Add centralized rate limiting service with Redis support

This commit introduces a new centralized rate limiting service that provides comprehensive rate limiting for all API endpoints. It includes user-based and IP-based limiting, configurable limits per endpoint and user tier, and supports burst protection. The service utilizes a Redis backend for distributed systems, ensuring efficient rate limit management across multiple instances. Additionally, it features methods for checking rate limits, retrieving current status, and resetting limits, enhancing the overall API usability and performance.

kid-sid committed:
Added offline testing suite for rate limiting service

This commit introduces a new script for offline testing of the rate limiting service, allowing verification of its functionality without a running server. The script includes tests for service initialization, health checks, rate limit configuration, middleware import, utility functions, and configuration import, providing comprehensive coverage of the rate limiting implementation.

kid-sid committed:
Added rate limiting test suite

This commit introduces a new test script for the rate limiting functionality, allowing for comprehensive testing of various scenarios including rate limit status, inference rate limiting, authentication rate limiting, and verification of rate limit headers. The script provides a structured approach to validate the behavior of the rate limiting service and ensures proper handling of rate limits across different endpoints.

kid-sid committed:
Added rate limiting utilities for FastAPI integration

This commit introduces a new module containing helper functions for integrating rate limiting into FastAPI routers. The utilities include functions to retrieve rate limit information, check limits for users and IPs, reset user limits, and format rate limit messages. Additionally, a dependency for obtaining the current user with rate limit checks is provided, enhancing the API's rate limiting capabilities and usability.

kid-sid committed:
Changed the file location.

kid-sid committed:
Added the doc strings.

kid-sid committed:
Changed the file location.

kid-sid committed:
Added a centralized rate limiting service to prevent API abuse.

kid-sid committed:
Added the config file for the rate limiter.

kid-sid committed:
Updated import path for the rate limiting service in rate_limiting_utils.py to reflect new directory structure.

kid-sid committed:
Updated PROJECT_STATUS.md to reflect the implementation of a comprehensive rate limiting system. Version updated to 2.4.0, highlighting the completion of critical security features, including multi-tier rate limits, Redis-based distributed limiting, and enhanced error handling. All endpoints are now protected, ensuring production readiness with full security measures in place.

kid-sid committed:
Enhanced README.md to reflect the implementation of a comprehensive rate limiting system, including multi-tier architecture and Redis-based support. Updated features section to highlight security improvements and added detailed rate limiting configuration tables. Clarified production readiness status and outlined key features of the rate limiting algorithm.

kid-sid committed:
Updated import path for the rate limiter in main.py to align with the new directory structure.

pinku3019 commented on middleware/rate_limiting_middleware.py:
```diff
@@ -0,0 +1,227 @@
+    ):
+        super().__init__(app)
+        self.bypass_endpoints = bypass_endpoints or [
+            "/health",
```
Repeat data for bypassing, make it in a single area for defaults and add new if required.

pinku3019 commented on middleware/rate_limiting_middleware.py:
```diff
@@ -0,0 +1,227 @@
+            "/static"
+        ]
+        self.bypass_patterns = bypass_patterns or [
+            "/static/",
```
Same repeated sections here

pinku3019 commented on middleware/rate_limiting_middleware.py:
```diff
@@ -0,0 +1,227 @@
+            return response
+            
+        except RateLimitExceeded as e:
+            logger.warning(f"Rate limit exceeded: {e.message}")
```
Create a custom dir/file for all type of custom and updating exceptions, and call them here any where required,
You can do it also in another story or PR

pinku3019 commented on routes/inference_router.py:
```diff
@@ -342,3 +343,35 @@ async def clear_cache():
+            - usage_percentages (dict): Usage as percentages
+    
+    Raises:
+        HTTPException: 401 if user is not authenticated
```
Its raising 500 error but this says 401

pinku3019 commented on services/rate_limiter/rate_limits_config.py:
```diff
@@ -0,0 +1,215 @@
+    
+    # Authentication endpoints
+    "auth": {
+        RateLimitTier.FREE: RateLimitConfig(
```
These valus should be stored separately and called for specific requirements with additional updates.

pinku3019 commented on utils/rate_limiting_utils.py:
```diff
@@ -0,0 +1,281 @@
+            "message": "Rate limit status retrieved successfully"
+        }
+        
+    except Exception as e:
```
Always try to use perticular exception , not the base class

pinku3019 commented on utils/rate_limiting_utils.py:
```diff
@@ -0,0 +1,281 @@
+    """
+    Get current authenticated user with rate limiting check
+    
+    This can be used as a dependency in FastAPI routes
```
Doc is not complete

kid-sid committed:
Added a PromptLoader utility for managing prompt files.
This commit introduces a new utility class, PromptLoader, for loading and managing prompt files. It includes features such as caching, error handling, and methods to list available prompts and check for their existence. Additionally, convenience functions are provided for easy access to specific prompts, enhancing the overall functionality of the prompt management system.

kid-sid committed:
Integrated lazy prompt loading into optimize_prompt function. This update adds functionality to load a system prompt from a file, enhancing the prompt optimization process by incorporating predefined prompts for improved clarity and effectiveness.

kid-sid committed:
Refactored optimize_prompt function to utilize loaded system prompt from file. This change enhances the prompt optimization process by replacing hardcoded content with a dynamic prompt, improving maintainability and clarity.

kid-sid committed:
Added lazy_prompt.txt to provide guidelines for improving user prompts. This new file outlines how to enhance prompt clarity and effectiveness while maintaining original intent.

kid-sid committed:
Added pro_prompt.txt to provide advanced strategies for transforming user prompts. This new file outlines techniques such as chain-of-thought reasoning, few-shot learning, and role-based prompting, aimed at enhancing clarity and effectiveness while preserving the original intent.

kid-sid committed:
Removed unnecessary prompt templetes.
